logstash-core/spec/support/shared_contexts.rb

```diff
@@ -34,28 +36,42 @@
 end
 
 shared_context "api setup" do
+
+  ##
+  # blocks until the condition returns true, or the limit has passed
+  # @return [true] if the condition was met
+  # @return [false] if the condition was NOT met
+  def block_until(limit_seconds, &condition)
+    deadline = Time.now + limit_seconds
+    loop.with_index do |_,try|
+      break if Time.now >= deadline
+      return true if condition.call
+
+      next_sleep = [(2.0**(try))/10, 2, deadline - Time.now].min
+      Kernel::sleep(next_sleep) unless next_sleep <= 0
+    end
+    # one last try
+    condition.call
+  end
+
   before :all do
     clear_data_dir
-    settings = mock_settings
-    config_string = "input { generator {id => 'api-generator-pipeline' count => 100 } } output { dummyoutput {} }"
-    settings.set("config.string", config_string)
-    settings.set("config.reload.automatic", false)
-    @agent = make_test_agent(settings)
-    @agent.execute
-    @pipelines_registry = LogStash::PipelinesRegistry.new
-    pipeline_config = mock_pipeline_config(:main, "input { generator { id => '123' } } output { null {} }")
-    pipeline_creator =  LogStash::PipelineAction::Create.new(pipeline_config, @agent.metric)
-    expect(pipeline_creator.execute(@agent, @pipelines_registry)).to be_truthy
-    pipeline_config = mock_pipeline_config(:secondary, "input { generator { id => '123' } } output { null {} }")
-    pipeline_creator =  LogStash::PipelineAction::Create.new(pipeline_config, @agent.metric)
-    expect(pipeline_creator.execute(@agent, @pipelines_registry)).to be_truthy
```

Comment:
this section started an agent with a single pipeline intentionally designed to complete itself and close with `config.reload.automatic: false` _so that_ `@agent.execute` would start the API but not block, and then side-channel created two running pipelines directly in a new pipeline registry that was _not_ connected to our agent, but whose spawned pipelines used the metrics store from our agent. The net effect is that the agent does not know about those two pipelines, and so our API is running in a state where its canonical truth is fractured between the agent and the metrics store.

By creating a minimal implementation of a source loader and using `config.reload.automatic: true`, we can trivially cause new pipelines to get loaded and reloaded _in the normal way_, which means that our metrics store and our agent share a common source-of-truth.